c2cgeoform/views/abstract_views.py

```diff
@@ -248,28 +246,45 @@ def edit(self):
         self._populate_widgets(form.schema)
         rendered = form.render(form.schema.dictify(obj), request=self._request)
         return {
-            'form': rendered,
-            'deform_dependencies': form.get_widget_resources()
+            'form': form,
+            'form_rendered': rendered,
+            'deform_dependencies': form.get_widget_resources(),
+            'duplicate_url': self._request.route_url(
+                'c2cgeoform_item_action',
+                id=self._request.matchdict.get('id'),
+                action='duplicate')
         }
 
-    def duplicate(self):
-        obj = self._new_object()
-        tmpl = self._get_object()
-
-        insp = inspect(self._model)
+    def copy_members_if_duplicates(self, model, source):
+        dest = model()
+        insp = inspect(model)
         for prop in insp.attrs:
             if isinstance(prop, ColumnProperty):
                 if model_attr_info(prop.columns[0], 'c2cgeoform', 'duplicate'):
-                    setattr(obj, prop.key, getattr(tmpl, prop.key))
+                    setattr(dest, prop.key, getattr(source, prop.key))
             if isinstance(prop, RelationshipProperty):
-                pass
+                if model_attr_info(prop, 'c2cgeoform', 'duplicate'):
+                    duplicate = [self.copy_members_if_duplicates(
```

Comment:
list depends on prop.uselist
note that you may have a column and a related relationship for foreign keys (in that case you get uselist = False, but the reverse assertion may be false, if uselist have been positioned manually in the model, for 1-1 relationship for example).